src/Grpc.StatusProto/README.md

```diff
@@ -0,0 +1,305 @@
+    Google.Rpc.Status? rpcStatus = ex.GetRpcStatus(); // Extension method
+
+    if (rpcStatus != null)
+    {
+        // Decode each message item in the details in turn
+        foreach (var msg in status.UnpackDetailMessage())
+        {
+            switch (msg)
+            {
+                case ErrorInfo errorInfo:
+                    Console.WriteLine($"ErrorInfo: Reason: {errorInfo.Reason}, Domain: {errorInfo.Domain}");
+                    foreach (var md in errorInfo.Metadata)
+                    {
+                        Console.WriteLine($"\tKey: {md.Key}, Value: {md.Value}");
+                    }
+                    break;
+
+                case BadRequest badRequest:
+                    Console.WriteLine("BadRequest:");
+                    foreach (BadRequest.Types.FieldViolation fv in badRequest.FieldViolations)
+                    {
+                        Console.WriteLine($"\tField: {fv.Field}, Description: {fv.Description}");
+                    }
+                    break;
+
+                // Other cases handled here ...
+            }
+        }
+    }
+
+```
+
+## Returning errors within gRPC streams
+
+The model described above allows you to return an error status when the gRPC call finishes.
+
+As an extension to the richer error model you may want to allow servers to send back
+multiple statuses when streaming responses without terminating the call.
+
+One way of doing this is to include a `google.rpc.Status` message in the definition
+of the response messages returned by the server.  The client should also be aware
+that it may receive a status in the response.
+
+For example:
+
+
+```protobuf
+service WidgetLookupProvider {
+    rpc streamingLookup(stream WidgetReq) returns (stream WidgetRsp) {}
+}
+
+message WidgetReq {
+    string widget_name = 1;
+}
+
+message WidgetRsp {
+    oneof message{
+        // details when ok
+        string widget_details = 1;
+        // or error details
+        google.rpc.Status status = 2;
+   }   
+}
+```
+
+Note: the  [status.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
+and [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
+files are not currently provided in any NuGet packages. If you wish to use these in
+your own message definitions then you will need to copy these into your own project.
+
+Example server code fragment:
+```C#
+ while (await requestStream.MoveNext())
+{
+    var request = requestStream.Current;
+    var response = new WidgetRsp();
+
+    // ... process the request ...
+
+    // to return an error
+    if (error)
+    {
+        response.Status = new Google.Rpc.Status { /* ... */ };
+    } else
+    {
+        response.WidgetDetails = "the details";
+    }
+}
+```
+
+Example client code fragment:
+```C#
+
+// reading the responses
+var responseReaderTask = Task.Run(async () =>
+{
+    while (await call.ResponseStream.MoveNext())
```

Comment:
Change this to use an await foreach + ReadAllAsync. It's a bit cleaner.

Example:

https://github.com/grpc/grpc-dotnet/blob/246c43d4574a5e2ba3fb4f397df304dfb5688b2d/examples/Vigor/Client/Program.cs#L35-L38